Fix z3 installation on PR CI jobs

thomasspriggs opened the pull request:
Includes updating z3 to 4.8.10 for Windows CI jobs. So that the version, which is currently most recently released is available for use in Windows regression tests. Z3 installation was previously broken, due to the chocolatey package being out of date / broken and due to the installation being missing from various jobs.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Add z3 to Ubuntu 18 package pull request check job

This was previously missing, which would cause any tests which needed Z3
to fail this job.

thomasspriggs committed:
Add z3 installation for windows + make job

This was previously missing, which would cause any tests which needed Z3
to fail this job.

thomasspriggs committed:
Log version of z3 installed in CI

For confirmation of version and that installation is successfull. This
supports easier debugging for both the github actions `.yaml` and for
future test failures. As a bonus we get fast failing CI jobs if the
dependency isn't installed properly, because it fails at this step
rather than during testing.

This z3 version check is done as a separate step because adding to
`$env:GITHUB_PATH` doesn't take effect the path until the next step.

thomasspriggs committed:
Use `Invoke-WebRequest` instead of `choco` to install `z3` on Windows

Because the `choco` installation of `z3` seems to be broken/outdated.

Using `Invoke-WebRequest` instead of `choco` allows us to pin the
version installed and to get a version which includes the `z3.exe`, not
just `z3.dll`. Using `nuget` was also considered, so that we were still
using a package manager. But the up to date `nuget` package was missing
the `.exe` file.

`wget` is supposed to be aliased to `Invoke-WebRequest` in powershell.
However `Invoke-WebRequest` is used rather than `wget` as `wget` does
not seem to work for windows github actions jobs.

NlightNFotis commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -287,7 +287,7 @@ jobs:
       - name: Fetch dependencies
         run: |
-          choco install winflexbison3 strawberryperl
+          choco install winflexbison3 strawberryperl z3
```
Should this commit be dropped?

thomasspriggs replied:
This commit exists, so that I am not doing both adding and updating in subsequent commits.

NlightNFotis reviewed (commented):
I have a couple of remarks:

1) This PR claims to be upgrading z3 on windows, but it also makes a number of other changes, such as probing for z3 version on mac and ubuntu jobs? Is this on purpose? If that's the case, can the PR be renamed to accurately reflect scope of changes, or submit the changes as another PR?
2) One of the CI jobs is failing because it's probing for `z3` when it's not installed.

TGWDB reviewed (changes requested):
A couple of bookkeeping changes:

1. The PR comment claims "latest version", which may be true today but will not be in future. Since the PR explicitly sets the version to 4.8.10 can we make that clear? (Specially since we could probably also use something like https://github.com/Z3Prover/z3/releases/latest for the true latest.)
2. Several places claim to "Check z3 version", but I think they only display the version, there does not appear to be a "check" that this is correct?

thomasspriggs commented:
After discussion with Fotis, I am going to change this PR to fix Z3 installation across all jobs and add a test. Please wait for me to push the extra changes, before reviewing.

thomasspriggs commented:
> A couple of bookkeeping changes:
> 
> 1. The PR comment claims "latest version", which may be true today but will not be in future. Since the PR explicitly sets the version to 4.8.10 can we make that clear? (Specially since we could probably also use something like https://github.com/Z3Prover/z3/releases/latest for the true latest.)

I think this should now be fixed.

> 2. Several places claim to "Check z3 version", but I think they only display the version, there does not appear to be a "check" that this is correct?

It is useful as is for confirming that Z3 is installed, in the path and logging the version. We could in theory check that the information printed is what we expect. But I think the changes are useful as is, without adding more functionality. I'll re-word the message a little, in order to try and make this clear.

thomasspriggs committed:
Add test that Z3 solver support is present and working

This confirms that Z3 is present and working in CI, as well as that
cbmc's support for Z3 is not entirely broken.

thomasspriggs commented:
@TGWDB Your comments should now be addressed. Please re-review when you have a minute.

SaswatPadhi commented:
Not a blocker if we don't have broken tests right now, but we might soon run into cases where a regression test might pass on Windows & Mac but failing on Ubuntu 18.04 or 20.04 because they're running older versions of Z3 (4.4.1 and 4.8.7 respectively).

Would it be possible to `wget` the https://github.com/Z3Prover/z3/releases/download/z3-4.8.10/z3-4.8.10-x64-ubuntu-18.04.zip binary from https://github.com/Z3Prover/z3/releases/tag/z3-4.8.10 and just move it to `/usr/local/bin` or some directory in `$PATH`?

@tautschnig is listed as the maintainer of the `z3` Ubuntu package (https://packages.ubuntu.com/bionic/z3), may be he could help too.

NlightNFotis reviewed (approved):
LGTM

thomasspriggs commented:
@SaswatPadhi Now that this is merged, you should be able to re-base and merge https://github.com/diffblue/cbmc/pull/5904
